### PR TITLE
Do not cleanup immutable workspace identity cache

### PR DESCRIPTION
**Pull request by wolfs**
to improve artifact transform performance for up-to-date
transforms.

### CODE REVIEW

**Comment by lptr** [comment]
I think we either need to keep `BuildOutputCleanupRegistry` alive between builds, too, or we need to make sure that we somehow record outputs even when we load from the identity cache.

**Comment by lptr** [comment]
...unless `BuildOutputCleanupRegistry` is about recording things created during the current build, in which case we are okay how things are organized now, but perhaps we could clarify what we mean by `isOutputOwnedByBuild()`.

If we choose to register outputs not created during the current build, then we should make sure that we also register outputs of up-to-date work, which we don't do currently.

**Comment by wolfs** [comment]
Why is `BuildOutputCleanupRegistry` related to re-using the `IdentityCache`? `BuildOutputCleanupRegistry` receives some safe to delete location by plugins. Currently, those plugins are the base plugin (`build` directories and anything cleaned by `clean`) and the java plugin (source set outputs). So tasks do not register anything there.

**Comment by lptr** [comment]
Sorry, I meant the `OutputFilesRepository` (which is used for cleanup only) — we should clarify these names.

**Comment by wolfs** [comment]
`OutputFilesRepository` is persisted to disk and is a cross-build cache already. It stores the locations produced by a build until the Gradle version changes. So there is also nothing to do.

**Comment by lptr** [comment]
👍 thanks for checking.

**Comment by lptr** [comment]
I think we should still clarify the names, though not necessarily in this PR.

**Comment by wolfs** [comment]
@bot-gradle test QFL

**Comment by wolfs** [comment]
@lptr PTAL!

File: subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java [hunk 1]
```diff
@@ -232,4 +242,52 @@ protected V maybeGetRetainedValue(Class<?> key) {
...
+    private static class BuildAgeEvictionCrossBuildInMemoryCache<K, V> implements CrossBuildInMemoryCache<K, V>, SessionLifecycleListener {
+        private final ManualEvictionInMemoryCache<K, V> delegate = new ManualEvictionInMemoryCache<>();
+        private final ConcurrentMap<K, Boolean> keysFromPreviousBuild = new ConcurrentHashMap<>();
+        private final ConcurrentMap<K, Boolean> keysFromCurrentBuild = new ConcurrentHashMap<>();
+
+        @Override
+        public V get(K key, Function<? super K, ? extends V> factory) {
+            keysFromCurrentBuild.put(key, Boolean.TRUE);
+            return delegate.get(key, factory);
+        }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java:255]: I understand why we don't need to override `Cache.get(K, Supplier)` here, but it seems a bit fragile... wonder if there's a better solution? Capturing the delegation in a `ForwardingCache` base class could be one potential way.

File: subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java [hunk 2]
```diff
@@ -232,4 +242,52 @@ protected V maybeGetRetainedValue(Class<?> key) {
...
+            keysToRetain.addAll(keysFromPreviousBuild.keySet());
+            keysToRetain.addAll(keysFromCurrentBuild.keySet());
+
+            delegate.retainAll(keysToRetain);
+
+            keysFromPreviousBuild.clear();
+            keysFromPreviousBuild.putAll(keysFromCurrentBuild);
+            keysFromCurrentBuild.clear();
+        }
+    }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java:292]: I don't see a test case for this where we actually test the eviction, only that some things are retained when they should be. I think we should add such a test.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformParallelIntegrationTest.groovy [hunk 3]
```diff
@@ -217,16 +217,16 @@ class ArtifactTransformParallelIntegrationTest extends AbstractDependencyResolut
...
         buildFile << """
             def a = file('a.jar')
-            a.text = '1234'
+            a.text = '12345'
             def b = file('b.jar')
-            b.text = '12'
+            b.text = '124'
             def c = file('c.jar')
-            c.text = '123'
+            c.text = '1236'
```
Comment (lptr) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformParallelIntegrationTest.groovy:229]: Why are these changes needed?

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/TestTransformationWorkspaceProvider.java [hunk 4]
```diff
@@ -42,7 +41,7 @@ public ExecutionHistoryStore getExecutionHistoryStore() {
 
     @Override
     public Cache<UnitOfWork.Identity, Try<ImmutableList<File>>> getIdentityCache() {
-        return emptyCache;
+        return new ManualEvictionInMemoryCache<>();
```
Comment (lptr) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/TestTransformationWorkspaceProvider.java:44]: Does this need to be created on every call to the getter?

File: subprojects/execution/build.gradle.kts [hunk 5]
```diff
@@ -26,10 +26,11 @@ dependencies {
     implementation(project(":files"))
     implementation(project(":snapshots"))
     implementation(project(":model-core"))
-    implementation(project(":persistent-cache"))
     implementation(project(":build-cache"))
     implementation(project(":build-cache-packaging"))
 
+    api(project(":persistent-cache"))
+
```
Comment (lptr) [inline subprojects/execution/build.gradle.kts:33]: `api` dependencies typically go above `implementation` ones.

But how come this is the first `api` dependency? We don't seem to need `files` for example. Shouldn't that be an `api` dependency? And some of the others seem like candidates, too...

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/ManualEvictionInMemoryCache.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+package org.gradle.cache;
+
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+
+public class ManualEvictionInMemoryCache<K, V> implements Cache<K, V> {
+    // Use 256 as initial size to start out with enough concurrency.
```
Comment (lptr) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/ManualEvictionInMemoryCache.java:25]: Why is 256 enough concurrency?

Comment (lptr) [reply subprojects/persistent-cache/src/main/java/org/gradle/cache/ManualEvictionInMemoryCache.java:25]: Passing an `int` to the constructor will set the concurrency level to `1`. Is this what we want?

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/CrossBuildInMemoryCacheFactory.java [hunk 7]
```diff
@@ -43,6 +43,15 @@
...
     <K, V> CrossBuildInMemoryCache<K, V> newCache();
 
+    /**
+     * Creates a new cache instance. Keys and values are always referenced using strong references.
+     *
+     * <p>Entries are only removed after each build session if they have not been used in this or the previous build.
+     *
+     * <p>Note: this should be used to create _only_ global/Gradle user home scoped instances.
+     */
+    <K, V> CrossBuildInMemoryCache<K, V> newCacheWithBuildAgedEviction();
```
Comment (lptr) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/CrossBuildInMemoryCacheFactory.java:53]: This name looks weird for me. What's a "build aged eviction"? I don't have a better name ready, but I think we should find one.

**Review by lptr — CHANGES_REQUESTED** [review]

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformParallelIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformParallelIntegrationTest.groovy:229]: If we don't add those changes, then the transforms won't run in parallel since the hash code of the identities is the same. `ConcurrentHashMap.computeIfAbsent()` is only concurrent on different buckets.

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/TestTransformationWorkspaceProvider.java (hunk 4 above)
Comment (wolfs) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/TestTransformationWorkspaceProvider.java:44]: It should be, since if it not created on every call, then it may not be empty, right?

File: subprojects/execution/build.gradle.kts (hunk 5 above)
Comment (wolfs) [reply subprojects/execution/build.gradle.kts:33]: I agree that some other dependencies also probably should be API dependencies. It looks like this hasn't been a problem up to now, i.e we have been adding the extra dependencies in the consumers manually. Do we want to continue doing that and move the dependency on `persistent-cache` back to implementation?

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/ManualEvictionInMemoryCache.java (hunk 6 above)
Comment (wolfs) [reply subprojects/persistent-cache/src/main/java/org/gradle/cache/ManualEvictionInMemoryCache.java:25]: It will set the concurrency level to `1` in the `this` call, though the primary constructor only uses concurrency level as a sizing hint and for nothing else, so it is ignored.

File: subprojects/execution/build.gradle.kts (hunk 5 above)
Comment (lptr) [reply subprojects/execution/build.gradle.kts:33]: I'd rather follow what we did before in this PR, and discuss with @jjohannes how we wanted to approach this on a more general level.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/ManualEvictionInMemoryCache.java (hunk 6 above)
Comment (wolfs) [reply subprojects/persistent-cache/src/main/java/org/gradle/cache/ManualEvictionInMemoryCache.java:25]: > Why is 256 enough concurrency?

256 makes the tests pass which assume the 6 artifact transforms to run in parallel. As soon as we add more elements, the size will increase. 256 seemed like a good default value to me. WDYT?

File: subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java:292]: I added a test case for the eviction.
